### ISSUE TITLE
Java 9 Support

### ISSUE THREAD

**Issue by 0x9090** [FasterXML/jackson-modules-base#37]
When compiling a Dropwizard project with Java 9, Jackson compilation throws the following error,

`WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.fasterxml.jackson.module.afterburner.util.MyClassLoader (file:/Users/user/Code/application.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int)
WARNING: Please consider reporting this to the maintainers of com.fasterxml.jackson.module.afterburner.util.MyClassLoader
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release`

**Comment by cowtowncoder** [comment]
Thank you. I have no idea what that means, or what could be done. But it's good to have here for future reference.

**Comment by 0x9090** [comment]
Looks like a common symptom is the use of internal packages. Could be what's going on here. It doesn't seem critical, but something to consider as more stuff moves to Java 9.

[https://blog.codefx.org/java/java-9-migration-guide/#Illegal-Access-To-Internal-APIs](https://blog.codefx.org/java/java-9-migration-guide/#Illegal-Access-To-Internal-APIs)

**Comment by black-snow** [comment]
Hitting this, too.

**Comment by istream-philion** [comment]
I've seeing this too.

It's related to the calls in loadAndResolve(ClassName className, byte[] byteCode) in MyClassLoader, specifically the defineClass override (I'm not sure which one, I'll see if testing catches it).

**Comment by cowtowncoder** [comment]
Ok. That's the "meat" of the system... so can not be ignored or skipped.

The part that is bit troubling is that it seems plausible fix could only be done in Java 9. And even Jackson 3.0 will only be Java 8 based.

Or maybe there is simpler way. I suspect if so, that would still need to go in Jackson 3.0, where bigger rewrites are possible.

**Comment by istream-philion** [comment]
Confirming my understanding: Jackson 3.0 will still be Java 8 based. 

Should I avoid deploying under Java 10, or can a Java 8 Jackson service fat jar run fine in a Java 10 Docker container?

**Comment by jsjohnst** [comment]
Adding this gets me past the warning...

`--add-opens java.base/java.lang=ALL-UNNAMED`

Still needs a better solution though, but its a start.

**Comment by cristian-mocanu-mob** [comment]
The problem is that Afterburner uses internal JDK API: the method ``java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int)`` is protected, and Oracle may simply removed it in even in a minor release.

Maybe Afterburner could use something else for bytecode generation, like ASM, JavaAssist, or Janino?

**Comment by cowtowncoder** [comment]
@cristian-mocanu-mob Afterburner uses ASM (for 2.x) and ByteBuddy (3.x). But those just generate bytecode, one still has to load the class in JVM. But if libraries offer help there, that'd be great.

I mean usage is literally just "load the class as per bytecode generate", there is no magic. I just haven't followed up on JVM developments in this area. It could be some simple change.

**Comment by jsjohnst** [comment]
@cowtowncoder Would something like this work (I’ve not done it, so I’m not sure)?

https://analyzejava.wordpress.com/2014/09/25/java-classloader-write-your-own-classloader/

**Comment by istream-philion** [comment]
I think the issue is relates to the block that starts at https://github.com/FasterXML/jackson-modules-base/blob/master/afterburner/src/main/java/com/fasterxml/jackson/module/afterburner/util/MyClassLoader.java#L81

Specifically, the warning calls out illegal reflection.

I don't understand the block at all. Why use reflection when just calling defineClass() should work?

Aside: I don't think it is the `protected` method that's the issue. That's a standard interface on a system class: even though is it protected, any class can subclass and invoke. It's not private, or an non-standard class (or even deprecated, there is an older defineClass).

**Comment by cowtowncoder** [comment]
@jsjohnst That article predates Java 9, so it wouldn't cover it. But actually as per @istream-philion there's bit more here; simple use case presented works but is not what module always does.

@istream-philion Ok so the comment suggests that this would be necessary for the case where class to access (or accessors) is not `public`. In those cases it is necessary to call `defineClass()` of the parent class loader of the value class being optimized: only that is able to access such members (and only when accessor class is in same package).
(it took a while to reverse-engineer this since `getParent()` could mean various other things...)

So: getting rid of this access would make it impossible to optimize access to any non-public field of any non-public class. And that would be quite unfortunate.

**Comment by cowtowncoder** [comment]
Ok: and just to clarify above... reflection is needed because `defineClass()` is `protected` and although sub-class can call it on `this`, it can not call that method on any other `ClassLoader` instance. And that's the crux. I can see how this goes against access restrictions Java 9's module system imposes, but if worst comes to worst, bytecode generation for accessors will be limited to public classes and their public members. And it is possible to configure module to do that already (see config in `AfterburnerModule`)

**Comment by cristian-mocanu-mob** [comment]
@cowtowncoder: am I correct that, to disable bytecode generation for private members, I would need to do ``AfterburnerModule.setUseValueClassLoader(false)``?

**Comment by cowtowncoder** [comment]
@cristian-mocanu-mob I would have to verify exact behavior here, but I think that `private` accessor are skipped in either case, and this setting would only affect handling of `protected` and "package protected" access cases.
This setting does affect whether Afterburner tries to access `ClassLoader` of value class itself, to access such non-public accessors, if (and only if) needed.

**Comment by cristian-mocanu-mob** [comment]
Setting ``AfterburnerModule.setUseValueClassLoader(false)`` fixed all warnings for me.
Since I am only mapping public getters/setters/constructors, this causes no issue for me.

**Comment by hc-codersatlas** [comment]
https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/MethodHandles.Lookup.html#defineClass-byte:A- is the replacement method (previously via unsafe)

**Comment by cowtowncoder** [comment]
@hc-codersatlas Thank you for suggesting this.

This will be a pain no matter what, since Jackson can not use Java 9 features directly with 2.x at all (ever). 3.x may at some point, but initially plan is only to require Java 8.

But as long as it's  just one method we can probably make it work dynamically, to allow Java 8 and Java 9+ both work with different implementation, dynamically loaded.

**Comment by stevenschlansker** [comment]
Have you considered targeting jdk9 or 10 for Jackson 3?  8 is going to be EOL very soon, and being 9-native would be a nice feature.

**Comment by cowtowncoder** [comment]
No. I don't use Java 9 myself, seems like more pain than value at this point. But when there is next long-term version (11?), that might change, within 3.x.

**Comment by hc-codersatlas** [comment]
11 is out in September and 3.x might not be out until then? i.e. might be worth considering.

**Comment by cristian-mocanu-mob** [comment]
@karussell: That should have worked.
I have the following, which is equivalent to what you have (because the setter returns ``this``):
```java
    @NonNull
    private static AfterburnerModule createAfterburnerModule() {
        final AfterburnerModule afterburnerModule = new AfterburnerModule();

        // make Afterburner generate bytecode only for public getters/setter and fields
        // without this, Java 9+ complains of "Illegal reflective access"
        afterburnerModule.setUseValueClassLoader(false);

        return afterburnerModule;
    }
```

Maybe your issue is not because of Afterburner.
Or maybe you use a different object mapper in the code that shows the warning.

**Comment by karussell** [comment]
Thanks, will investigate!

**Comment by karussell** [comment]
When I do not load this module at all I was able to remove the warning:

```java
class SomeApp extends Application {
 public void initialize(Bootstrap<GraphHopperServerConfiguration> bootstrap) {
   bootstrap.setObjectMapper(Jackson.newMinimalObjectMapper());// this does not load Afterburner at all
   ...
 }
}
```

(but I get a new warning regarding: Illegal reflective access by com.fasterxml.jackson.databind.util.ClassUtil to constructor java.util.Optional())

**Comment by cowtowncoder** [comment]
@karussell I would be interest in thing about `Optional` -- it should not be needed, at least if `jackson-datatype-jdk8` is included, which uses static factory method. However, if that is not included, databind probably accesses 0-arg constructor just like for any POJO.

**Comment by karussell** [comment]
I've tried to include this but without success. Maybe I did something wrong? Should I create a different issue to avoid "spamming" here or we can also discuss here: https://github.com/graphhopper/graphhopper/issues/1440

**Comment by stevenschlansker** [comment]
@karussell I believe the trace you pasted in that issue is different.  This specifically is related to defining classes, not deserializing Optional types.  I recommended a possible fix (make sure `jackson-datatype-jdk8` is installed properly everywhere, as Tatu mentions above, but I wish to reiterate that this is 99% likely the cause of your problem!) and if that is not the issue then you may consider opening a new ticket with reproduction steps.

**Comment by karussell** [comment]
Thanks for the update. I tried a bit around and finally I get no warnings with:

```
bootstrap.setObjectMapper(Jackson.newMinimalObjectMapper().registerModule(new Jdk8Module()));
```

For the following I still get the defineClass warning:

```
bootstrap.setObjectMapper(Jackson.newObjectMapper().registerModule(new AfterburnerModule().setUseValueClassLoader(false)));
```

Trying to use `Jackson.newMinimalObjectMapper()`:

```
bootstrap.setObjectMapper(Jackson.newMinimalObjectMapper().registerModule(new AfterburnerModule().setUseValueClassLoader(false)).registerModule(new Jdk8Module()));
```

does not print a warning but fails with:

```
java.lang.IllegalAccessError: class com.graphhopper.json.geo.JsonFeatureCollection$Access4JacksonDeserializer2d000ec4 tried to access field com.graphhopper.json.geo.JsonFeatureCollection.features (com.graphhopper.json.geo.JsonFeatureCollection$Access4JacksonDeserializer2d000ec4 is in unnamed module of loader com.fasterxml.jackson.module.afterburner.util.MyClassLoader @302c971f; com.graphhopper.json.geo.JsonFeatureCollection is in unnamed module of loader 'app')
```

**Comment by stevenschlansker** [comment]
A possible solution for JDK9+ is for Afterburner to accept a "private lookup" object for each module it is expected to synthesize accessors in.  It could then use that flavor of `defineClass` and only fall back to cracking `Unsafe` if it does not have a method handle.

**Comment by stevenschlansker** [comment]
Anyone feeling very tired of seeing these errors and very brave to try something new? 🐱

https://github.com/stevenschlansker/jackson-blackbird

**Comment by hc-codersatlas** [comment]
@stevenschlansker I recently compared this LambdaMetaFactory approach vs ReflectAsm and Unsafe and whilst the performance of all 3 is largely similar, the difference is the warmup time for Lambda is huge and so is the compilation time. The other 2 solutions are likely the way to go. Considering ReflectAsm and LambdaMetaFactory are both ways to dynamically generating an inner class, doing 1 per field/accessor has way too much overhead vs 1 per class.

**Comment by stevenschlansker** [comment]
Hi @hc-codersatlas , do you have a good benchmark that shows the warmup time being unacceptable?
For example, in my rough draft of a benchmark, run over 10 JVM forks in single-shot startup time mode:
https://gist.github.com/stevenschlansker/c1ec1e4bd3346c5811cbc74f5bd6e995

For the single shot time, it's definitely sometimes slow, but not appreciably slower than Afterburner.  For the types of applications I use even a full 50ms difference is lost in the ~5s total startup time.

To look at compile time, I ran it for 10 minute and captured hs_comp profile:
https://gist.github.com/stevenschlansker/38a9cf0a15988d7058d377fee558a831

In this case, you can see that the "vanilla" definitely does less compiling, but again Afterburner and Blackbird are within margin of error.

I also tested in a moderately complex (~20 pojo types) web application and noticed no startup time regression, and a slightly better response time for large data payloads after warmup.

I understand your point about custom class writing, but it is definitely a lot more complicated to do -- just look at the masses of `asm` code I got to delete by doing it this way -- and so far every benchmark I try shows it compares just fine.  Do you have a counterexample or a suggestion for how to improve my benchmarks?

**Comment by hc-codersatlas** [comment]
@stevenschlansker https://www.optaplanner.org/blog/2018/01/09/JavaReflectionButMuchFaster.html for example. Lambda vs say reflection is 30x for bootstrap. This test was only for a small number of pojo. I tried something similar with a production app that had a lot more pojos and the times were a lot higher. In regards to the benchmark, are you benchmarking the setup time? I do notice that it's all in the setup method and not the actual benchmark runs.

I'm not sure if it's more complicated. This has already been resolved by e.g. https://github.com/EsotericSoftware/reflectasm or Unsafe (which is still valid as of Java 12). These are much better options for various reasons. The Lambda method boils down to compiling inner classes and this fills up the MetaSpace, code cache etc too, i.e. there are other side effects.

There's also the issue of references: how are you holding onto these functions? Web apps that do class loading / unloading e.g. Java EE stuff would break or create memory leaks potentially.

**Comment by stevenschlansker** [comment]
You're right that my initial pass at startup time wasn't right.  I did a better job here:

https://github.com/stevenschlansker/jackson-blackbird/pull/1

Now I attempt to measure a single-shot startup that measures the cost of ObjectMapper initialization, module registration, and then serializes a list of three different bean types once.

Vanilla unexpectedly is the cheapest.  As of jdk11, `afterburner` is a smidge faster, but by jdk12 `blackbird` starts winning.  In either case it's only a few milliseconds either way, which in the context of a server, is not much.

I appreciate the reference to the optaplanner blog.  Unless I'm mistaken I think the comparison is a bit apples to oranges: they are testing on jdk8 (I am targeting 11+, since we use `privateLookupIn` which was only added in 9), and the startup time comparisons are interesting but don't compare the Afterburner approach (writing your own classfile and loading it) at all.  Their cited bootstrap costs are measured in microseconds, so while 30x looks scary, the biggest value (excluding JavaCompiler) is 40 µs.  So again it should fall into the noise.  (Our minimal services still take a good 3-4 seconds to bootstrap, so that's what I'm using as my baseline)

`reflectasm` similarly looks interesting but their project has benchmarks as of jdk7 (!) and the users group looks all but dead: https://groups.google.com/forum/#!forum/reflectasm-users

Regarding references and unloading, the generated lambdas are held as normal fields, for example:
https://github.com/stevenschlansker/jackson-blackbird/blob/master/blackbird/src/main/java/com/fasterxml/jackson/module/blackbird/deser/SettableLongProperty.java#L14

Barring a leak in the JVM itself, when your `war` is unloaded (and presumably Jackson with it) it should all be collected normally.
Of course if you do crazy things like load Jackson from your Tomcat server's parent classloader and store your ObjectMapper in a static field, there's a chance of references leaking.  But I think most sane uses shouldn't have any troubles.

My hope is that the JDK authors have given us a supported and seemingly reasonably performant way to get optimized performance with no external libraries.  We should have the advantage that, as the implementation underlying such core features as lambda and method references, its performance will be under the microscope (and has been - you can already see improvements in every release since introduction!)

Do you think the new benchmarks are better?  Did you try it against an actual application and note specific problems I could track down?  Thanks for any additional input.

**Comment by hc-codersatlas** [comment]
@stevenschlansker maybe I don't understand or never understood the fear of Unsafe. There were rumors, discussions and issues going from Java 8->9 and to date (as of Java 12 and 13?) it's still available. It's not getting removed. It's not being marked as deprecated. There are no class generation or other overheads and performance is comparable. There are some specific edge case caveats that apply to loops and vectorization but I don't see it being an issue here. For this use case (afterburner) I still see it as the best solution, i.e. to answer your question it is a supported and performance oriented way. Most of the internal jdk libraries itself use it. 

There s a place for LambdaFactory, but I don't think it's for fields but methods. ReflectAsm is used in Kryo and other projects. It just works and there are generally no issues. In terms of Enterprise and related technology, Hibernate and anyone else all use a similar method whether via JavaAssist, ByteBuddy or similar to generate a single class (vs the many in lambdaFactory) per class (1:1 mapping) to achieve whatever is required here. Imagine every developer ignores this and starts generating 100s of these classes. This is what makes Groovy, Scala, etc require a lot more memory and slows things down.  Java already gives a bad impression of being memory hungry - this is why. It doesn't have to be.

It's not so much if 30x is scary or not - why do it when there's a better way? Having said that - it is scary when you have an actual large project and do development on it, e.g. Spring has Devtools and that auto restarts on changes (literally).

**Comment by stevenschlansker** [comment]
@hc-codersatlas, I wouldn't characterize it as _fear_ so much.  I understand it's not imminently targeted for removal.  It does look like the existing approach could be adapted to the new Lookup defineClass.  But there's other reasons to explore alternatives to dynamic classloading.  For example, Graal's `substratevm` (native image generation) has [support for translating lambda expressions](https://github.com/oracle/graal/blob/master/substratevm/LIMITATIONS.md#lambda-expressions) so in principle, we could eventually generate native code in an AOT compiler!
(caveat: I have not tried this yet!)

I do agree: the current approach works best when you have (non-private) accessor methods.  Personally, I think that's good style anyway: if the property is intended to be accessed outside the class, it *is* part of the API, and `setAccessible` has always been kinda icky anyway.

It won't be for everyone, especially at the start.  But "exploding" the call sites, while it may be bad for your use case, is good for others: it lets the optimizer do extremely aggressive inlining that wouldn't otherwise be possible.

I've got a [PR open contributing it as a Jackson module](https://github.com/FasterXML/jackson-modules-base/pull/85).  Code reviews or even just "it worked for me" reports appreciated!

**Comment by alibenmessaoud** [comment]
Same warning messages. I use Dropwizard 1.3.12 on Java 11.

**Comment by Sounie** [comment]
I wonder if multi JVM version support could be smoothly introduced by applying  
[Multi-Release JAR Files](http://openjdk.java.net/jeps/238)

Keep the "root" code as compiled for Java 8, and introduce an alternative implementation for the ClassLoader reflection for Java 9+ to pick up.

It will probably introduce some more complexity to the build process, but this seems to be a good match for what the JEP is intended to facilitate.

--
Stephen

> @hc-codersatlas Thank you for suggesting this.
> 
> This will be a pain no matter what, since Jackson can not use Java 9 features directly with 2.x at all (ever). 3.x may at some point, but initially plan is only to require Java 8.
> 
> But as long as it's just one method we can probably make it work dynamically, to allow Java 8 and Java 9+ both work with different implementation, dynamically loaded.

**Comment by OrangeDog** [comment]
> I have no idea what that means

In case it wasn't clear yet - Afterburner is using reflection to access a non-public API outside of its own module (which in this case is "all non-module code"), that may behave differently (or in general, be missing) in any given JDK implementation (across both vendor and version).

While it does seem unlikely that a `protected` method would break, there is no guarantee that something bad wouldn't happen if this was called from outside a properly-conforming subclass of `ClassLoader`.

**Comment by cowtowncoder** [comment]
At this point discussion on this issue, and its original vague wording mean that it is difficult to follow.
Reporting yet more "hey I run DropWizard and see a warning" will not help; and due to changes in module-status of other libraries (i.e. which of one's dependencies are full modules, use automatic module name, or none of above) change symtoms.
So basically this is not really actionable at this point.

So: I think new, more specific issue(s) should be filed, with possible reference to this issue as background/context.

I will probably close this one soon.

**Comment by stevenschlansker** [comment]
One possible route of exploration to maintain current Afterburner functionality is to change from `Unsafe.defineClass` to `MethodHandles.Lookup.defineClass`.  This has the possibility of preserving the existing ability to "crack open" non-public types.

**Comment by cowtowncoder** [comment]
@stevenschlansker There are couple of other issues: would that approach fall under them? Or a new issue? Since `MethodHandles` is available on JDK 7 that seems like possibility (plus I would be ok with Afterburner going Java 8 even during Jackson 2.x).

**Comment by stevenschlansker** [comment]
I haven't tried!  Just leaving it as a suggestion for follow-up work in case anyone really wants to update Afterburner proper.  We are happily running the Blackbird pre-release and have not looked back :)

**Comment by cowtowncoder** [comment]
@stevenschlansker ah right. One more question (which I may have asked previously but forgot answer to...) : does each `MethodHandle` typically generate a single class instance, similar to how a lambda would work?

**Comment by cogman** [comment]
@cowtowncoder I didn't see that you got a response here (Don't know if it came offline).  But it is my understanding that no new classes are created for method references.

The source is complex around it, but from my cursory glance, it does not look like any classes are being introduced.

AFAIK, exact behavior isn't specified (so YMMV with other JVMs)

**Comment by cogman** [comment]
Also of note, if you are willing to bump the requirement up to Java 9, then you can use VarHandle to do to object fields what you are doing to object methods.  I don't know if that is out of the question, though, since there appears to be a lot of people out there stuck on 8.

**Comment by cowtowncoder** [comment]
If no classes are loaded in memory, that'd be a significant plus I think.

O Java 9, that would not be an option for 2.x, but by the time 3.0 is out, perhaps 11 or 14 (one of LTS versions) could become baseline. At least for Afterburner (although whether one can use post-Java8 JDK for generating later versions and java 8 in same repo I don't know -- Oracle has not been very supportive of doing java 8 builds from later versions AFAIK).

**Comment by wrprice** [comment]
Please pardon my lurking, and I hope I understand the concern correctly...

> although whether one can use post-Java8 JDK for generating later versions and java 8 in same repo I don't know

OpenJDK 11 `javac` with the `--release 8` flag will emit JDK8-compatible bytecode just fine.  I've been doing it w/ a production codebase for years.  The trickier part is modifying your build scripts to make multi-release JAR files.  You'd build the majority of the project using javac 11 and `--release 8`, then "layer" Java 11 classes on top with a second javac step (without `--release`) that builds alternate versions of only those classes that want to use JDK9+ features.  The APIs of the alternate versions should match the Java 8 versions, but the implementation details can differ.  Place the output of the first compilation step on the classpath for the second step so that the Java 11 alternates properly link to any shared/common Java 8 classes.

Then run the test suite twice, once each with JDK8 and JDK11.

**Comment by cowtowncoder** [comment]
Right, I don't think I'd really want to develop or support true multi-release jars in that sense (although it is good that is possible), but rather isolate support for later JDK versions into distinct modules.
But it may be necessary to build different maven sub-projects with different JDK settings, which should be doable.

**Comment by OrangeDog** [comment]
Given that blackbird is now a core module, is there any intention to fix afterburner, or should users just switch?

**Comment by cowtowncoder** [comment]
If someone can suggest or contribute a fix I'd be happy to merge it, but I am not aware of one.
Given this, yes, it probably makes sense for post-Java-8 users to have a look at Blackbird. My only concern there is that there has been less real world usage, but there is only one way to get over that.

**Comment by eperret** [comment]
Is this still an issue in versions of Java later than 9?
